### PR TITLE
Improve Pedersen hash efficiency by using unsafe add

### PR DESCRIPTION
**Pull request by imeckler**
This PR shaves 1 constraint off of the curve addition used in the Pedersen hash. Whereas before hashing cost (2 + 4) constraints per triple, it will now cost (2 + 3) constraints per triple. Concretely this should be about a 12% improvement in hashing inside the SNARK.

Doing this though requires introducing a scary function. That function is `add_unsafe` which is a checked function for adding two curve points. The scary part is in `div_unsafe`. Which is a checked function for dividing one field element by another. In the event that both field elements are zero however, the result is totally unconstrained and could be chosen by the prover to be arbitrary.

This then causes `add_unsafe` to have the following behavior. 

- If `p = - q` then `add_unsafe p q` makes an unsatisfiable assertion.
- If `p = q` then `add_unsafe p q` could be many different curve points(!)
- If `p !=  +/- q`, then `add_unsafe p q` is genuinely the curve point `p + q`

Thus, `add_unsafe` must never be called on two potentially identical curve points. All uses of it are in the context of Pedersen hashing in which we add curve points of the form

```
add_unsafe (a_i * g_i + ... + a_j + g_j)  (a_k * g_k + ... + a_l + g_l)
```

with the intervals `[i, j]` and `[k, l]` being disjoint and
where the `g_r` are the random pedersen parameters and `a_s in {+/- 1, …, +/- 4}`.

I claim that the usual argument indicates that one should not be able to find `a_s` causing those two sums to be identical unless one can compute discrete logs in `G`.

### CODE REVIEW

**Comment by cmr** [comment]
in theory I like this but it seems to be broken.

File: src/lib/snarky/src/curves.ml [hunk 1]
```diff
@@ -661,6 +664,36 @@ module Make_weierstrass_checked
       in
       (cx, cy))
 
+  (* This function MUST NOT be called UNLESS you are certain the two points
```
Comment (bkase) [inline src/lib/snarky/src/curves.ml]: can you make this a doc-comment on the signature?

File: src/lib/snarky/src/curves.ml [hunk 2]
```diff
@@ -661,6 +664,36 @@ module Make_weierstrass_checked
...
+         x <> 0, y = 0 -> 0 * z = x. (i.e., unsat)
+
+         x <> 0, y <> 0 -> z = x / y
+      *)
+      let%map () = assert_r1cs y z x in
+      z
+    in
+    fun p q ->
+      let%map r = add' ~div:div_unsafe p q in
+      `I_thought_about_this_very_carefully r
```
Comment (bkase) [inline src/lib/snarky/src/curves.ml]: I love this

File: src/lib/snarky/src/curves.ml [hunk 3]
```diff
@@ -661,6 +664,36 @@ module Make_weierstrass_checked
...
 
+  (* This function MUST NOT be called UNLESS you are certain the two points
+   on which it is called are not equal. If it is called on equal points,
+   the prover can return almost any curve point they want to from this function. *)
+  let add_unsafe =
+    let open Let_syntax in
+    let div_unsafe x y =
+      let%bind z =
+        provide_witness Field.typ
+          As_prover.(map2 (read_var x) (read_var x) ~f:Field.Infix.( / ))
```
[the remark this answers is not in the document]
Comment (cmr) [reply src/lib/snarky/src/curves.ml]: d'oh indeed.

**Pull request merged**